cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,326 @@
+rad install kubernetes --set global.aws.IRSA.enabled=true #default false
+
+if enabled, it should print a success/ failure message after updating the pod spec to mount the token. It should also print a information message to the complete of configuration by using rad credential register and rad env update.
+
+rad credential register aws --type <IRSA> --role <roleARN>
+rad credential register aws --type <access-key> --access-key-id <access-key-id> --secret-access-key <secret-access-key>
+
+rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 
+
+### interactive rad init support
+```
+% rad init --full  
+Enter an environment name 
+>default 
+
+% rad init --full  
+Select your cloud provider                     
+1. Azure                                  
+> 2. AWS                                     
+3.[back] 
+
+% rad init --full  
+
+Select the identity for the AWS cloud provider 
+1. access key                              
+> 2. IRSA  
+
+% rad init --full  
+
+Enter the AWS RoleARN for configuring IRSA: <role-arn-radius>
+
+```
+
+
+Once radius is installed, we use rad env update to update the aws provider config
+```
+% rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 
+```
+
+**Sample Output:**
+
+n/a
+
+**Sample Recipe Contract:**
+
+n/a
+
+### CLI Design
+
+`rad init --full` will need to be updated to ask the user to choose between AWS access keys and IRSA. rad credential register aws should now support two credential types, and associated parameters.
+
+We should update the Radius Helm chart to allow the user to enable IRSA during installation with the `rad install kubernets --set global.aws.IRSA.enabled=true` value. This must be passed to ucp and rp charts through values and when the value is set the pod spec should mount the service-account token.
+
+### Multi-tenancy
+
+Currently, we support only one credential. However, for future, we might want to support multiple credentials potentially across AWS accounts ( say, one for QA team and other for Dev). The code changes for IRSA should be compatible with these changes in future. Since IRSA setup establishes trust between the cluster - oidc provider and AWS, once we support multiple credentials, we should be able to assume different roleARN based on which environment we are deploying from. We just have to establish right trust entities in all roles.  
+
+### Implementation Details
+
+#### UCP
+
+UCP is responsible for communicating with AWS to deploy AWS resources. 
+UCP should be able to assume the role specified in the credential and use this to communicate with AWS since it has the service-token mounted.
+
+#### Bicep
+
+NA
+
+#### Deployment Engine
+
+NA
+
+#### Core RP  / Recipes RP
+
+Terraform provider requires AWS credentials in order to deploy recipes.
+
+RP should be able to assume the role specified in the credential and use this to communicate with AWS since its pod will have the service-token mounted.
```

Comment:
I might have misunderstood this so just want to confirm - Currently RP fetches the credentials stored by UCP, were you thinking of switching from that to RP directly fetching the credentials by doing the assuming role instead?

Reply:
Thank you for bringing this up. I have not looked into recipe / rp implementation of credential handling. There is no reason for RP to change the way it fetches credentials. It would continue to fetch it through UCP, but now, since it gets a role-arn as credential, it should be updated to be able to use that.